10-query/01b-pattern-negation.md

```diff
@@ -0,0 +1,334 @@
+like this:
+
+```
+$x isa person;
+not { 
+    ($x, father: $y) isa parentship;
+    not { ($y) isa employment; };
+};
+```
+
+Please note, nesting of negation blocks is only allowed in queries, but not in rules.
+
+## Negation blocks: DOs and DONTs
+
+Please note, the following restrictions apply to negation blocks:
+- for each negation block in a query, at least one variable in the negation block must be bound to a statement outside of the negation block
+This ensures that set difference operations are performed on sets that are not disjoint.
+- variables in negation blocks are local to the block they are defined in
+- only conjunctive statements are allowed
+
+
+## Negation in rules
+As we have already mentioned, we can use negation blocks within rules.
+ 
+However, when inserting negation blocks in rules, currently the following restrictions apply:
+- all restrictions applying to queries with negation blocks
+- each rule can only have a single negation block
+- nested negation blocks are not supported
+- recursion with negation blocks is not supported 
+
+We will illustrate the use of negation with rules with a graphical example.
+
+Let us define a network of nodes with possible links between nodes:
+
+```
+traversable sub entity,
+    has index;
+    plays from,
+    plays to;
+
+node sub traversable;
+
+link sub relation, relates from, relates to;
+```
+
+Then we can define two nodes as being reachable if there exists a link between them:
+
+```
+reachable sub relation, relates from, relates to;
+reachabilityA sube rule,
+when {
+    (from: $x, to: $y) isa link;
+},
+then {
+    (from: $x, to: $y) isa reachable;
+};
+
+reachabilityB sub rule,
+when {
+    (from: $x, to: $z) isa link;
+    (from: $z, to: $y) isa reachable;
+},
+then {
+    (from: $x, to: $y) isa reachable;
+};
+```
+
+Consequently, with the use of negation we can define links that are indirect:
+
+```
+indirect-link sub relation, relates from, relates to;
+indirect-link-rule sub rule,
+when {
+    (from: $x, to: $y) isa reachable;
+    not {(from: $x, to: $y) isa link;};
+},
+then {
+    (from: $x, to: $y) isa indirect-link;
+};
+```
+
+or, mark the unreachable nodes by specifying the following rule:
+
+```
+unreachable sub relation, relates from, relates to;
+unreachability-rule
+when {
+    $x isa node;
+    $y isa node;
+    not {(from: $x, to: $y) isa reachable;};
+},
+then {
+    (from: $x, to: $y) isa unreachable;
+};
+```
+
+Please note the explicit addition of the `node` types in the body of the rule. This is to to ensure the boundedness condition
+is satisfied as well as to maintain the expected meaning of `unreachable` relation.
+
+
+
+## Monotonicity of reasoning with negation
+
+In this subsection we will address an important property of reasoning being the monotonicity property.
+We say that reasoning is monotonic if previously derived facts remain true upon addition of new knowledge. Up to this point Graql reasoning was monotonic. However with
```

Comment:
Let's change this a bit because these docs may stand when negation isn't a recent addition. I also want us to make clear that Graql still is monotonic if you don't use negation in rules.

_We say that reasoning is monotonic if previously derived facts remain true upon addition of new knowledge. Without using pattern negation, Graql reasoning is monotonic. When pattern negation is used in a rule this property no longer holds._

Let's add that this is a point to be aware of, not a constraint. Maybe something like:

_Conserving monotonicity is not required, but it is important to be aware that it is being sacrificed when using pattern negation in rules._
_Let us illustrate the difference with a popular example:_